Disallow empty arbitrary values

RobinMalfait opened the pull request:
This PR makes the candidate parser more strict by not allowing empty arbitrary values.

Examples that are not allowed anymore:

- `bg-[]` — arbitrary value
- `bg-()` — arbitrary value, var shorthand
- `bg-[length:]` — arbitrary value, with typehint
- `bg-(length:)` — arbitrary value, with typehint, var shorthand
- `bg-red-500/[]` — arbitrary modifier
- `bg-red-500/()` — arbitrary modifier, var shorthand
- `data-[]:flex` — arbitrary value for variant
- `data-():flex` — arbitrary value for variant, var shorthand
- `group-visible/[]:flex` — arbitrary modifier for variant
- `group-visible/():flex` — arbitrary modifier for variant, var shorthand

If you are trying to trick the parser by injecting some spaces like this:

- `bg-[_]`

Then that is also not allowed.

RobinMalfait committed:
add failing test

RobinMalfait committed:
ensure empty arbitrary values are invalid

This is implemented for various places:

- `bg-[]` — arbitrary value
- `bg-()` — arbitrary value, var shorthand
- `bg-[length:]` — arbitrary value, with typehint
- `bg-(length:)` — arbitrary value, with typehint, var shorthand
- `bg-red-500/[]` — arbitrary modifier
- `bg-red-500/()` — arbitrary modifier, var shorthand
- `data-[]:flex` — arbitrary value for variant
- `data-():flex` — arbitrary value for variant, var shorthand
- `group-visible/[]:flex` — arbitrary modifier for variant
- `group-visible/():flex` — arbitrary modifier for variant, var shorthand

RobinMalfait committed:
move parsing of the modifier up

Which allows us to delete 2 places where we parsed the modifier

RobinMalfait committed:
add test index

RobinMalfait committed:
ensure migrating invalid candidates with empty `[]`  work

RobinMalfait committed:
prepare and cleanup raw candidates

This is used in 2 spots:

1. When handling prefixes
2. When modernizing values

The issues this tries to solve is to ensure that valid v3 candidates but
invalid v4 candidates are migrated correctly. E.g.: `group-[]:flex` is
valid in v3, but not in v4 (it won't parse). With this workaround, it is
parsing correctly.

RobinMalfait committed:
add more test cases

RobinMalfait committed:
update changelog

philipp-spiess commented on packages/@tailwindcss-upgrade/src/template/codemods/modernize-arbitrary-values.test.ts:
```diff
@@ -31,6 +31,14 @@ test.each([
   ['group-[]:flex', 'in-[.group]:flex'],
   ['group-[]/name:flex', 'in-[.group\\/name]:flex'],
 
+  // There is no good equivalent for this one right now:
+  ['group-hover/[]:flex', 'group-hover/[]:flex'],
```
The comment here is confusing in relation to this PR since the latter won't parse anymore so are we actually breaking this with no fallback? 🤔

RobinMalfait replied:
This is technically a breaking change because it did parse in v3, but as far as I can tell nobody uses that (because it doesn't really make sense)

This is what it produces in v3 today:
```css
.group\/\[\]:hover .group-hover\/\[\]\:flex {
  display: flex;
}
```

https://sourcegraph.com/search?q=context:global+/%5B%5D+repohasfile:tailwind.config&patternType=keyword&sm=0

RobinMalfait commented on packages/tailwindcss/src/candidate.test.ts:
```diff
@@ -1416,3 +1416,130 @@ it('should parse candidates with a prefix', () => {
+  // Empty arbitrary value
+  'bg-[]',
+  'bg-()',
+  //  Tricking the parser with a space is not allowed
```
Hah I did that on purpose because it's part of the `Empty arbitrary value` section 😅

RobinMalfait committed:
fix comment

React brain… this should be about variants…

RobinMalfait commented on packages/tailwindcss/src/candidate.test.ts:
```diff
@@ -1416,3 +1416,130 @@ it('should parse candidates with a prefix', () => {
+  'bg-[_]/(_)',
+  'bg-(_)/(_)',
+
+  // Functional components
```
React brain... this should say `Functional variants` and the other one `Compound variants` because I wanted to verify that we handled functional variants and compound variants correctly (since they are different types and handled separately).

RobinMalfait committed:
add `handleEmptyArbitraryValues` codemod

This adds a dedicated codemod that can transform `peer-[]:flex` to
`peer-[&]:flex` and `group-[]:flex` to `group-[&]:flex`.

This ensures that the variants are parsing and generate an equivalent in
v4. Later codemods also ensure that `group-[&]:flex` get converted to
`in-[.group]:flex` instead.

RobinMalfait committed:
allow `group-[]` in the Oxide scanner

This is only required for codemods right now

RobinMalfait committed:
add integration test

RobinMalfait committed:
update comments to look less weird while stil indenting
